### ISSUE TITLE
Can create session, but can't get back into it.

### ISSUE THREAD

**Issue by Billybangleballs** [rdesktop/rdesktop#342]
I upgraded to **rdesktop:armhf 1.8.6-2\~deb9u1** from **rdesktop:armhf 1.8.4-1\~deb9u1** (which worked fine btw), and now, although I can start a session on the remote computer, if I disconnect for any reason, I can't reconnect. I get an error msg :-

```
ERROR: rdp.c:140: rdp_recv(), unexpected stream overrun0000 03 00 00 1e 02 f0 80 68 00 01 03 eb 70 10 08 00 .......h....p...
0010 02 03 c7 3e 63 ee 58 8c 45 d9 04 00 16 00       ...>c.X.E.....
```

I also can't connect to existing sessions started by other clients, I get the login screen, but it bombs out as soon as I put in the password.

If I could unapt-get upgrade, I would, but I'm not that technical.
Is this something that might be fixable, or have I broken it completely?

**Comment by Billybangleballs** [comment]
So after much googling, I figured out how to downgrade back to where it worked, but obviously this hasn't fixed the problem with rdesktop:armhf 1.8.6-2\~deb9u1 but it has fixed my immediate problem.

**Comment by CendioOssman** [comment]
Can you share the entire output from rdesktop with us? It's could be a duplicate of #320.

**Comment by Billybangleballs** [comment]
I'm sure the only output was the error message above, but I am unable to check this because I have downgraded back to a version that works.

**Comment by CendioOssman** [comment]
It will be difficult for us to debug this in that case. I'll close this for now and let's hope fixing #320 will also fix the issue you've seen.

**Comment by bernhardu** [comment]
There got Debian bug [934897](https://bugs.debian.org/934897) opened and I tried to look into it.
I guess it might describe the same issue as in this issue.
Attached is a file containing some more information and a backtrace of that location and a
git bisect pointing to commit 82fce18, but that might have just uncovered it.
There is also a patch avoiding the issue, if the end of the buffer is already reached.
Can you confirm my findings? @CendioOssman

**Comment by CendioOssman** [comment]
I can confirm that the issue was probably already there, just hidden since we didn't have any sanity checks.

I can't say if the patch is correct or not. Have you looked at Microsoft's protocol documentation and see what it says about this?

**Comment by bernhardu** [comment]
So does this justify to reopen this issue?

In version 50 or 35 (unfortunately no older version available as far as I see)
this section seems related:
2.2.8.1.1.1.1   Share Control Header (TS_SHARECONTROLHEADER)

As far as I see it is not mentioned that the `pduSource` field could be ommitted.
Otherwise it looks like XP sends already just a length of 0x4 bytes.

```
0000 03 00 00 1e 02 f0 80 68 00 01 03 eb 70 10 08 00 .......h....p...
0010 33 00 9a c6 db b8 94 e6 77 7e 04 00 16 00       3.......w~....
                                   04 00
                                   ^^^^^
                                   in_uint16_le(rdp_s, length);
                                   length == 0x4

                                         16 00
                                         ^^^^^
                                         in_uint16_le(rdp_s, pdu_type);
                                         pdu_type == 0x16 --> "& 0xf" --> 0x6 --> PDUTYPE_DEACTIVATEALLPDU
                                                  == 0b10110 --> & 0xfffffff0 >> 4 --> 0x1 --> TS_PROTOCOL_VERSION

                                               ??
                                               ^^ (in rdesktop 1.8.6)
                                               in_uint8s(rdp_s, 2);	/* userid */

                                               ?? ??
                                               ^^^^^ (in git master)
                                               in_uint16(s, pdu_source);	/* pduSource */
```

**Comment by CendioOssman** [comment]
That is very helpful, thanks.

The package makes complete sense, up until the point where `pduSource` source is omitted. A bug in XP I guess that all previous clients just ignored.

`pduSource` isn't actually used for anything, so I guess we could be tolerant of it being omitted. Could you try this patch:

```diff
diff --git a/rdp.c b/rdp.c
index 8ac0d28..a43a600 100644
--- a/rdp.c
+++ b/rdp.c
@@ -124,12 +124,22 @@ rdp_ts_in_share_control_header(STREAM s, uint8 * type, uint16 * length)
        }
 
        in_uint16_le(s, pdu_type);      /* pduType */
-       in_uint16(s, pdu_source);       /* pduSource */
-
        *type = pdu_type & 0xf;
 
+       /* XP omits pduSource for PDUTYPE_DEACTIVATEALLPDU for some reason */
+       if (*length == 4) {
+               logger(Protocol, Debug,
+                      "rdp_ts_in_share_control_header(), missing pduSource field for 0x%x PDU",
+                      *type);
+       } else {
+               in_uint16(s, pdu_source);       /* pduSource */
+       }
+
        /* Give just the size of the data */
-       *length -= 6;
+       if (*length >= 6)
+               *length -= 6;
+       else
+               *length = 0;
 
        return True;
 }
```

**Comment by bernhardu** [comment]
Unfortunately my tests until today were just for the 1.8.x branch.

In the master branch there got another patch 90fd660 introduced, that makes the connection to Windows XP impossible. If I revert that one, and put your proposed patch on top I am able to connect. Please find details in my tree mentioned above.

Unfortunately I could not make any clue what got wrong with 90fd660, therefore my test to revert it. Should that go into a different issue?

I guess issue 346 discusses also the 1.8.x branch because the PR is from a branch named 1.8.6_xp, and looks similar to my proposed downstream debian [patch](https://bugs.debian.org/cgi-bin/bugreport.cgi?att=2;bug=934897;filename=ignore-userid.diff;msg=10).

**Comment by CendioOssman** [comment]
Great. That really helps. I've opened issue #348 for this second problem. It's just confusing to mix things even if both affect Windows XP.

**Comment by CendioOssman** [comment]
Fix is in a646a5cce2ece3acc07ccead1e5f971cacc0db81.
